File: src/net/conn.c
#include <core.h>
#include <curl/curl.h>
#include <net/net.h>
#include <stdio.h>

constexpr auto ConnPollRate = Time_Millisecond * 10;

// Blocking mode:
//  Retry until output buffer is filled or there is an error.
static size_t NET_ConnRead(void* userdata,
                       void* ptr,
                       size_t size,
                       SDL_IOStatus* status) {
  Conn* conn = userdata;
  Uint8* out = ptr;
  size_t total = 0;

  *status = SDL_IO_STATUS_READY;

  while (total < size) {
    auto n = NET_CurlReadFromSocket(conn->sock, out + total, size - total);

    if (n < 0) {
      *status = SDL_IO_STATUS_ERROR;
      return total;
    }

    if (n == 0) {
      if (conn->Blocking) {
        Time_Sleep(ConnPollRate);
        continue;  // retry
      }

      if (total == 0)  // there is no data.
        *status = SDL_IO_STATUS_NOT_READY;
      return total;
    }

    total += n;

    if (!conn->Blocking)
      return total;
  }

  return total;
}
// Blocking mode:
//  Block untill all bytes are written or there is an error.
static size_t NET_ConnWrite(void* userdata,
                        const void* ptr,
                        size_t size,
                        SDL_IOStatus* status) {
  Conn* conn = userdata;
  const Uint8* in = ptr;
  size_t total = 0;

  *status = SDL_IO_STATUS_READY;

  while (total < size) {
    auto n = NET_CurlWriteToSocket(conn->sock, in + total, size - total);

    if (n < 0) {
      *status = SDL_IO_STATUS_ERROR;
      return total;
    }

    if (n == 0) {
      if (conn->Blocking) {
        Time_Sleep(ConnPollRate);
        continue;
      }

      *status = SDL_IO_STATUS_NOT_READY;
      return total;
    }

    total += n;

    if (!conn->Blocking)
      return total;
  }

  return total;
}

static bool NET_ConnClose(void* userdata) {
  Time_Sleep(Time_Second); // wait for all data to be sent.
  Conn* conn = userdata;
  curl_easy_cleanup(conn->sock);
  delete (conn);
  return true;
}

ConnDialResult NET_ConnDial(EM* em, string hostname) {
  EM* scratch = em_create_nested(em, 2048);
  defer {
    em_destroy(scratch);
  }
  // Allocate interface
  SDL_IOStreamInterface i;
  SDL_INIT_INTERFACE(&i);
  i.read = NET_ConnRead;
  i.write = NET_ConnWrite;
  i.close = NET_ConnClose;
  // Open TCP socket using curl.
  hostname = strCat(scratch, str("http://"), hostname);
  auto curl = NET_CurlCreateSocket(hostname);
  if (!curl.ok) {
    return Err(ConnDial, curl.err);
  }
  // return the Conn.
  auto conn = new (Conn);
  conn->sock = curl.result;
  conn->stream = SDL_OpenIO(&i, conn);
  return Ok(ConnDial, conn);
};

File: src/net/conn.go
package net

import (
	"betamine/net/curl"
	"betamine/sdl"

	"solod.dev/so/errors"
	"solod.dev/so/time"
)

var ErrResolveFailed = errors.New("Could not resolve hostname.")
var ErrConnectionFailed = errors.New("TCP connection failed.")
var ErrConnectionClosed = errors.New("Connection closed.")

// Read blocks until bytes are read or there is an error.
// It does not guarantee that the full buffer is used.
func (conn *Conn) Read(b []byte) (int, error) {
	if conn.closed { // already errored.
		return 0, ErrConnectionClosed
	}

	if len(b) == 0 {
		return 0, nil
	}
	for {
		n, err := curl.ReadFromSocket(conn.sock, &b[0], len(b))
		if err != nil {
			// conn.Close()
			return 0, err
		}
		if n != 0 {
			return n, nil
		}
		sdl.Delay(time.Millisecond * 50)
	}
}

// Write blocks until all bytes from the buffer have been written.
func (conn *Conn) Write(b []byte) (int, error) {
	if conn.closed {
		return 0, ErrConnectionClosed
	}

	total := 0
	for total < len(b) {
		n, err := curl.WriteToSocket(conn.sock, &b[total], len(b)-total)
		if err != nil {
			conn.Close()
			return total, err
		}

		if n == 0 {
			// avoid tight spin
			sdl.Delay(time.Millisecond * 50)
			continue
		}

		total += n
	}

	return total, nil
}

// Free Closes the connection and frees memory.
func (conn *Conn) Close() {
	if conn.closed {
		return
	}
	conn.closed = true
	curl.CloseSocket(conn.sock)
}

// Conn is a TCP client connection.
type Conn struct {
	closed bool
	sock   *curl.CURL
}
type ConnResult struct {
	val Conn
	err error
}

// Dial dials the connection with a default timeout.
func Dial(host string) (Conn, error) {
	host = "http://" + host
	conn := Conn{}
	err := curl.CreateSocket(host, &conn.sock)
	return conn, err
}
